File: R/lista05.R
# Carrega o setup de forma robusta, com ou sem run_all.R
if (is.null(getOption("econ_r_dir")) || is.null(getOption("econ_root"))) {
  .locate_r_dir_local <- function() {
    of <- tryCatch(sys.frame(1)$ofile, error = function(e) NA_character_)
    if (!is.null(of) && length(of) == 1 && is.character(of) && !is.na(of) && nzchar(of)) {
      return(dirname(normalizePath(of, winslash = "/", mustWork = FALSE)))
    }
    cwd <- normalizePath(getwd(), winslash = "/", mustWork = TRUE)
    if (basename(cwd) == "R" && file.exists(file.path(cwd, "00_setup.R"))) return(cwd)
    if (file.exists(file.path(cwd, "R", "00_setup.R"))) return(file.path(cwd, "R"))
    stop("Não foi possível localizar a pasta R do pacote.")
  }
  source(file.path(.locate_r_dir_local(), "00_setup.R"), chdir = TRUE)
} else {
  source(file.path(getOption("econ_r_dir"), "00_setup.R"), chdir = TRUE)
}

library(lmtest)
library(sandwich)

# Q2, se a base municipal for fornecida
mun_path <- file.path(DATA, 'municipios_lista05.csv')
if (file.exists(mun_path)) {
  mun <- read.csv(mun_path)
  mod <- lm(pobreza ~ no + ne + su + co + pibpc + escol + escol:no + escol:ne + escol:su + escol:co, data=mun)
  capture.output(summary(mod), file=file.path(TABLES, 'R_lista05_q02_ols.txt'))
  capture.output(coeftest(mod, vcov.=vcovHC(mod, type='HC1')), file=file.path(TABLES, 'R_lista05_q02_robusto.txt'))
  save_plot('R_lista05_q02_pobreza_observado_ajustado', {
    plot(mun$pobreza, fitted(mod), xlab='Pobreza observada', ylab='Pobreza ajustada', main='Pobreza observada versus ajustada')
    abline(0, 1)
  })
} else {
  writeLines('Arquivo municipios_lista05.csv não encontrado.', file.path(TABLES, 'R_lista05_q02_aviso.txt'))
}

# Q3
het <- read.csv(file.path(DATA, 'lista05_q03_hetero.csv'))
mod_het <- lm(y ~ x, data=het)
capture.output(summary(mod_het), file=file.path(TABLES, 'R_lista05_q03_ols.txt'))
capture.output(bptest(mod_het, ~ x + I(x^2), data=het), file=file.path(TABLES, 'R_lista05_q03_white.txt'))
capture.output(coeftest(mod_het, vcov.=vcovHC(mod_het, type='HC1')), file=file.path(TABLES, 'R_lista05_q03_robusto.txt'))
mod_wls <- lm(y ~ x, data=het, weights=1/(x^2))
capture.output(summary(mod_wls), file=file.path(TABLES, 'R_lista05_q03_wls.txt'))
save_plot('R_lista05_q03_hetero_scatter_mqo', {
  plot(het$x, het$y, xlab='Número médio de empregados', ylab='Salário médio por trabalhador', main='MQO e heterocedasticidade')
  abline(mod_het)
})
save_plot('R_lista05_q03_residuos_quadrados', {
  plot(het$x, resid(mod_het)^2, xlab='Número médio de empregados', ylab='Resíduo ao quadrado', main='Resíduos ao quadrado')
})

# Q5
X <- matrix(1, nrow=5, ncol=1)
y <- matrix(c(0,7,5,0,0), ncol=1)
beta <- solve(t(X) %*% X) %*% t(X) %*% y
write.csv(data.frame(parametro='beta_hat', valor=as.numeric(beta)), file.path(TABLES, 'R_lista05_q05.csv'), row.names=FALSE)
save_plot('R_lista05_q05_variancias_condicionais', {
  barplot(c(1,3,4,6,8), names.arg=1:5, xlab='Observação', ylab='Var(u_i | X)', main='Variância condicional do erro')
})
cat('Lista 5 R concluída.\n')

File: python/lista05.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.stats.diagnostic import het_white
from common import DATA, TABLES, add_const, savefig

# Q2 - base municipal, se fornecida
mun_path = DATA / 'municipios_lista05.csv'
if mun_path.exists():
    mun = pd.read_csv(mun_path)
    formula = 'pobreza ~ no + ne + su + co + pibpc + escol + escol:no + escol:ne + escol:su + escol:co'
    mod = smf.ols(formula, data=mun).fit()
    (TABLES / 'lista05_q02_ols.txt').write_text(mod.summary().as_text(), encoding='utf-8')
    rob = mod.get_robustcov_results(cov_type='HC1')
    (TABLES / 'lista05_q02_robusto.txt').write_text(rob.summary().as_text(), encoding='utf-8')

    mun['ajustado'] = mod.fittedvalues
    plt.figure(figsize=(7, 4))
    plt.scatter(mun['pobreza'], mun['ajustado'])
    lims = [min(mun['pobreza'].min(), mun['ajustado'].min()), max(mun['pobreza'].max(), mun['ajustado'].max())]
    plt.plot(lims, lims)
    plt.xlabel('Pobreza observada')
    plt.ylabel('Pobreza ajustada')
    plt.title('Lista 5 - Questão 2: observado versus ajustado')
    savefig(plt, 'lista05_q02_pobreza_observado_ajustado')
else:
    (TABLES / 'lista05_q02_aviso.txt').write_text('Arquivo data/manual/municipios_lista05.csv não encontrado; questão resolvida simbolicamente no notebook.', encoding='utf-8')

# Q3 - heterocedasticidade com tabela do enunciado
het = pd.read_csv(DATA / 'lista05_q03_hetero.csv')
ols = sm.OLS(het['y'], add_const(het['x'])).fit()
(TABLES / 'lista05_q03_ols.txt').write_text(ols.summary().as_text(), encoding='utf-8')
lm, lm_p, fstat, f_p = het_white(ols.resid, add_const(het['x']))
pd.DataFrame({'estatistica':['LM_White','p_LM','F_White','p_F'], 'valor':[lm,lm_p,fstat,f_p]}).to_csv(TABLES / 'lista05_q03_white.csv', index=False)
rob = ols.get_robustcov_results(cov_type='HC1')
(TABLES / 'lista05_q03_robusto.txt').write_text(rob.summary().as_text(), encoding='utf-8')
wls = sm.WLS(het['y'], add_const(het['x']), weights=1/(het['x']**2)).fit()
(TABLES / 'lista05_q03_wls.txt').write_text(wls.summary().as_text(), encoding='utf-8')

x_grid = np.linspace(het['x'].min(), het['x'].max(), 200)
y_grid = ols.params['const'] + ols.params['x']*x_grid
plt.figure(figsize=(7, 4))
plt.scatter(het['x'], het['y'], label='Observações')
plt.plot(x_grid, y_grid, label='Reta MQO')
plt.xlabel('Número médio de empregados')
plt.ylabel('Salário médio por trabalhador')
plt.title('Lista 5 - Questão 3: MQO')
plt.legend()
savefig(plt, 'lista05_q03_hetero_scatter_mqo')

plt.figure(figsize=(7, 4))
plt.scatter(het['x'], ols.resid**2)
plt.xlabel('Número médio de empregados')
plt.ylabel('Resíduo ao quadrado')
plt.title('Lista 5 - Questão 3: resíduos ao quadrado')
savefig(plt, 'lista05_q03_residuos_quadrados')

# Q5
X = np.ones((5,1))
y = np.array([0,7,5,0,0]).reshape(-1,1)
beta = np.linalg.inv(X.T @ X) @ X.T @ y
pd.DataFrame({'parametro':['beta_hat'], 'valor':[float(beta[0,0])]}).to_csv(TABLES / 'lista05_q05.csv', index=False)
variancias = np.array([1,3,4,6,8])
plt.figure(figsize=(7, 4))
plt.bar(np.arange(1,6), variancias)
plt.xlabel('Observação')
plt.ylabel('Var(u_i | X)')
plt.title('Lista 5 - Questão 5: variância condicional do erro')
savefig(plt, 'lista05_q05_variancias_condicionais')

print('Lista 5 concluída.')
